### ISSUE TITLE
consumerGroup status always show consuming partitions "STOPPED"

### ISSUE THREAD

**Issue by tapanhalani** [linkedin/Burrow#237]
{
      "topic": "lyearn.populatedb.user",
      "partition": 9,
      "status": "STOP",
      "start": {
        "offset": 7952519,
        "timestamp": 1499084499648,
        "lag": 2107,
        "max_offset": 7952527
      },
      "end": {
        "offset": 7952734,
        "timestamp": 1499084523970,
        "lag": 1892,
        "max_offset": 7952741
      }
    }

In the above partition object, I am getting "STOP" partition status. But when I check the lag using yahoo kafka manager as well as kafka cli, I can see the messages being consumed by this partition. This behaviour happens everytime above 2000 messages comes in the partition, causing periodic mails from burrow giving consumer group error status, until all the messages are not consumed completely. 

I am using no-kafka client library in node.js to write consumers, and manually commit the offset using library function after processing on the message received is completed, instead of automatic periodic commits. The lagcheck interval is default 10. Why is this happening?

Also, I can't understand at what intervals in time, will the offset be sampled in the lag-check window? I mean in the example given here, it says every 1 minute due to auto commit offset interval. But since I am using manual commit offset with every message, how to visualise this use-case?

**Comment by johnmbw** [comment]
We've been seeing this quite a bit too, but always for one out of 80 consumers.  The topic being consumed by that consumer is lower volume than the others, but still gets a few messages a second at least.

Also under normal operation burrow seems to use 20% CPU, but when the problem occurs it suddenly moves to 120% CPU (so guessing something is looping somewhere and not updating offsets for that consumer).

Looking at the code for Rule 4:
```go
// Rule 4 - Offsets haven't been committed in a while
if ((time.Now().Unix() * 1000) - lastOffset.Timestamp) > (lastOffset.Timestamp - firstOffset.Timestamp) {
	status.Status = protocol.StatusError
	thispart.Status = protocol.StatusStop
	status.Partitions = append(status.Partitions, thispart)
	continue
}
```

It feels like it might not take much for that to alert falsely - even if lag is actually low/zero - as it runs before a lot of the other rules.

Once burrow gets into this state restarting seems to fix it.

**Comment by johnmbw** [comment]
After further examination it appears that the goroutine processing the rules never receives the offsets for the problem topics (enable trace-level in `offsets_store` to confirm that).

I've confirmed that the problem topic was actually committing offsets though.  My suspicion is that the CPU spike (possibly some other problem) causes the call to `timeoutSendOffset`:

https://github.com/linkedin/Burrow/blob/dce021af8389c83c1389c232ecbdbe12f122e56b/kafka_client.go#L360

to drop offsets in the floor and this will disproportionally affect quieter topics.

At this point I've hit my limit on knowledge of Golang, so not quite sure how this could be fixed.

**Comment by toddpalino** [comment]
@tapanhalani, what you're running into is probably a complication from committing offsets so frequently. If you have the number of stored offsets set to 10, and you rapidly commit 10 offsets in 10 seconds, and then there's 11 seconds before the next offset commit, you're going to get reported as stopped.

You could also hit this if all the offset commits are inside the min-distance config, which is the minimum amount of time between stored offset commits. If min-distance is set to 1 second, and you commit 10 offsets inside that second, only the first one will be stored. This will result in bad behavior as well. This is something we can resolve by updating the last offset commit with the new values if it's inside min-distance, rather than dropping it.

The other issue would need to be solved by rethinking what a stopped partition is. The current definition in burrow is "offsets have not been committed for longer than the interval between the first offset we have stored and the last one". It assumes a somewhat regular offset commit interval.

**Comment by toddpalino** [comment]
Resolved by 1.0 rewrite
